x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/ChangePoint.java

```diff
@@ -0,0 +1,207 @@
+        this.key = key;
+        this.targetType = targetType;
+        this.targetPvalue = targetPvalue;
+    }
+
+    private ChangePoint(StreamInput in) throws IOException {
+        this(
+            Source.readFrom((PlanStreamInput) in),
+            in.readNamedWriteable(LogicalPlan.class),
+            in.readNamedWriteable(NamedExpression.class),
+            in.readNamedWriteable(NamedExpression.class),
+            in.readNamedWriteable(Attribute.class),
+            in.readNamedWriteable(Attribute.class)
+        );
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        throw new UnsupportedOperationException("ChangePoint should run on the coordinating node, so never be serialized.");
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    protected NodeInfo<ChangePoint> info() {
+        return NodeInfo.create(this, ChangePoint::new, child(), value, key, targetType, targetPvalue);
+    }
+
+    @Override
+    public UnaryPlan replaceChild(LogicalPlan newChild) {
+        return new ChangePoint(source(), newChild, value, key, targetType, targetPvalue);
+    }
+
+    @Override
+    public List<Attribute> output() {
+        if (output == null) {
+            output = NamedExpressions.mergeOutputAttributes(List.of(targetType, targetPvalue), child().output());
+        }
+        return output;
+    }
+
+    public NamedExpression value() {
+        return value;
+    }
+
+    public NamedExpression key() {
+        return key;
+    }
+
+    public Attribute targetType() {
+        return targetType;
+    }
+
+    public Attribute targetPvalue() {
+        return targetPvalue;
+    }
+
+    @Override
+    public List<Attribute> generatedAttributes() {
+        return List.of(targetType, targetPvalue);
+    }
+
+    @Override
+    public ChangePoint withGeneratedNames(List<String> newNames) {
+        checkNumberOfNewNames(newNames);
+        Attribute newTargetType;
+        if (targetType.name().equals(newNames.get(0))) {
+            newTargetType = targetType;
+        } else {
+            newTargetType = targetType.withName(newNames.get(0)).withId(new NameId());
+        }
+        Attribute newTargetPvalue;
+        if (targetPvalue.name().equals(newNames.get(1))) {
+            newTargetPvalue = targetPvalue;
+        } else {
+            newTargetPvalue = targetPvalue.withName(newNames.get(1)).withId(new NameId());
+        }
+        return new ChangePoint(source(), child(), value, key, newTargetType, newTargetPvalue);
+    }
+
+    @Override
+    protected AttributeSet computeReferences() {
+        return Expressions.references(List.of(key, value));
+    }
+
+    @Override
+    public boolean expressionsResolved() {
+        return value.resolved() && key.resolved();
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(value, key, targetType, targetPvalue, child());
```

Comment:
nit: The usual pattern is `return Objects.hash(super.hashCode(), value, key, targetType, targetPvalue)`.